tests/test-cross-context-dispatch/ccd-ee10-webapp/src/main/java/org/eclipse/jetty/tests/ccd/ee10/CCDServlet.java

```diff
@@ -0,0 +1,108 @@
+
+        dispatchPlan.addEvent("%s.service() dispatcherType=%s method=%s requestUri=%s",
+            this.getClass().getName(),
+            req.getDispatcherType(), req.getMethod(), req.getRequestURI());
+
+        Step step;
+
+        while ((step = dispatchPlan.popStep()) != null)
+        {
+            if (step instanceof Step.ContextRedispatch contextRedispatchStep)
+            {
+                ServletContext otherContext = getServletContext().getContext(contextRedispatchStep.getContextPath());
+                if (otherContext == null)
+                    throw new NullPointerException("ServletContext.getContext(\"" + contextRedispatchStep.getContextPath() + "\") returned null");
+                RequestDispatcher dispatcher = otherContext.getRequestDispatcher(contextRedispatchStep.getDispatchPath());
+                if (dispatcher == null)
+                    throw new NullPointerException("ServletContext.getRequestDispatcher(\"" + contextRedispatchStep.getDispatchPath() + "\") returned null");
+                switch (contextRedispatchStep.getDispatchType())
+                {
+                    case FORWARD -> dispatcher.forward(req, resp);
+                    case INCLUDE -> dispatcher.include(req, resp);
+                }
+                return;
+            }
+            else if (step instanceof Step.RequestDispatch requestDispatchStep)
+            {
+                RequestDispatcher dispatcher = req.getRequestDispatcher(requestDispatchStep.getDispatchPath());
+                if (dispatcher == null)
+                    throw new NullPointerException("HttpServletRequest.getRequestDispatcher(\"" + requestDispatchStep.getDispatchPath() + "\") returned null");
+                switch (requestDispatchStep.getDispatchType())
+                {
+                    case FORWARD -> dispatcher.forward(req, resp);
+                    case INCLUDE -> dispatcher.include(req, resp);
+                }
+                return;
+            }
+            else if (step instanceof Step.GetHttpSession getHttpSessionTask)
+            {
+                HttpSession session = req.getSession(false);
+                if (session == null)
+                {
+                    dispatchPlan.addEvent("%s.service() HttpSession is null",
+                        this.getClass().getName());
+                }
+                else
+                {
+                    String name = getHttpSessionTask.getName();
+                    Object value = session.getAttribute(name);
+                    dispatchPlan.addEvent("%s.service() HttpSession exists: [%s]=[%s]",
+                        this.getClass().getName(),
+                        name,
+                        Objects.toString(value)
+                    );
+                }
+
+            }
+            else if (step instanceof Step.HttpSessionSetAttribute sessionSetAttribute)
+            {
+                HttpSession session = req.getSession(true);
```

Comment:
I'm putting this comment on the `ee10`  `CCDServlet`, but it goes for all the others as well. I think each step  needs to generate a `dispatchPlan.addEvent`. I was just following through the event output from one of the tests and I got confused as it looked like there was a `"HttpSession exists: null"` where it shouldn't be, but then I realized that the `HttpSetSessionAttribute` step doesn't generate an event output. I think its important for consistency and comprehensibility that every step generates at least 1 event output.

For extra credit, it might be nice to put the generation of those events inside the `Step` itself, so each servlet doesn't haven't to do so much boilerplate code (with the possibility of making errors), so something like:

```java
interface Step
{
   void addEvent(DispatchPlan plan, String... args);
}


class GetHttpSession extends AbstractStep
{

    public void addEvent(DispatchPlan plan, String... args)
    {
        plan.addEvent("%s.service() HttpSession exists: [%s]=[%s]",
                        args[0],
                        args[1],
                        args[2]
    }
}
```

so then the `CCDServlet`s code becomes:

```java
   step.addEvent(
                        dispatchPlan,
                        this.getClass().getName(),
                        name,
                        Objects.toString(value));
```